libs/algo/odc/algo/_tiff.py

```diff
@@ -0,0 +1,518 @@
+        return self.__str__()
+
+    @property
+    def name(self) -> str:
+        return self._out.name
+
+    @property
+    def info(self) -> GeoRasterInfo:
+        return self._info
+
+    def close(self):
+        self._out.close()
+
+    def __del__(self):
+        self.close()
+
+        if self._mem_mine:
+            self._mem_mine.close()
+            self._mem_mine = None
+
+    def __setitem__(self, key: NumpyIndex, item: np.ndarray):
+        ndim = len(key)
+        info = self._info
+        assert ndim in (2, 3)
+
+        yx_key = key[info.axis : info.axis + 2]
+        if ndim == 2:
+            assert info.axis == 0
+            assert item.ndim == 2
+            bands: Union[int, Tuple[int, ...]] = 1
+            block = item
+        elif ndim == 3:
+            if info.axis == 0:
+                # Y, X, B
+                bands = np_slice_to_idx(key[2], info.count)
+                if item.ndim == 2:
+                    block = np.expand_dims(item, axis=0)
+                else:
+                    # rio expects band to be the first dimension
+                    block = item.transpose([2, 0, 1])
+            else:
+                # B, Y, X
+                bands = np_slice_to_idx(key[0], info.count)
+                if item.ndim == 2:
+                    block = np.expand_dims(item, axis=0)
+                else:
+                    block = item
+
+            # rio wants 1 based indexing
+            bands = tuple(i + 1 for i in bands)
+        else:
+            raise ValueError("Only accept 2 and 3 dimensional data")
+
+        win = Window.from_slices(*yx_key, height=info.height, width=info.width)
+        if self._lock:
+            with self._lock:
+                self._out.write(block, indexes=bands, window=win)
+        else:
+            self._out.write(block, indexes=bands, window=win)
+
+
+class COGSink:
+    def __init__(
+        self,
+        info: GeoRasterInfo,
+        dst: str,
+        blocksize: Optional[int] = None,
+        ovr_blocksize: Optional[int] = None,
+        bigtiff: Union[bool, str] = "auto",
+        lock: bool = True,
+        temp_folder: Optional[str] = None,
+        overview_resampling: str = "average",
+        rio_opts_first_pass: Optional[Dict[str, Any]] = None,
+        use_final_blocksizes: bool = False,
+        **extra_rio_opts,
+    ):
+        if blocksize is None:
+            blocksize = 512
+
+        if ovr_blocksize is None:
+            ovr_blocksize = blocksize
+
+        if bigtiff == "auto":
+            # do bigtiff if raw raster is larger than 4GB
+            bigtiff = info.raster_size() > (1 << 32)
+
+        opts = dict(
+            driver="GTiff",
+            bigtiff=bigtiff,
+            tiled=True,
+            blockxsize=_adjust_blocksize(blocksize, info.width),
+            blockysize=_adjust_blocksize(blocksize, info.height),
+            compress="DEFLATE",
+            zlevel=6,
+            predictor=2,
+            num_threads="ALL_CPUS",
```

Comment:
probably another naive question, but does this use the all the cpus on the machine? or all the cpus available to the dask worker?

Reply:
That I'm not sure, it is probably using all the cores that are visible, so it could be problematic inside k8s where process is allocated fewer cores via resource limits.